airflow/dags/puhti_download_demo.py

```diff
@@ -0,0 +1,121 @@
+"""
+#### DAG for downloading METS and ALTO files for a single binding to a remote server (Puhti).
+"""
+
+from datetime import timedelta
+
+from airflow import DAG
+from airflow.operators.dummy_operator import DummyOperator
+from airflow.operators.python import PythonOperator
+from airflow.providers.http.sensors.http import HttpSensor
+from airflow.contrib.hooks.ssh_hook import SSHHook
+from airflow.models import Connection
+from airflow import settings
+
+from harvester.file import ALTOFile
+from harvester.mets import METS
+from harvester.pmh_interface import PMH_API
+
+DC_IDENTIFIER = "https://digi.kansalliskirjasto.fi/sanomalehti/binding/379973"
+BASE_PATH = "/scratch/project_2006633/nlf-harvester/downloads"
+
+
+default_args = {
+    "owner": "Kielipankki",
+    "start_date": "2022-10-01",
+    "retries": 5,
+    "retry_delay": timedelta(seconds=10),
+}
+
+
+def create_nlf_conn(conn_id):
+    session = settings.Session()
+    conn_ids = [conn.conn_id for conn in session.query(Connection).all()]
+    if not conn_id in conn_ids:
+        conn = Connection(
+            conn_id=conn_id,
+            conn_type="HTTP",
+            host="digi.kansalliskirjasto.fi/interfaces/OAI-PMH",
+            schema="HTTPS",
+        )
+        session.add(conn)
+        session.commit()
+
+
+def save_mets_for_id(ssh_conn_id):
+    api = PMH_API(url="https://digi.kansalliskirjasto.fi/interfaces/OAI-PMH")
+    ssh_hook = SSHHook(ssh_conn_id=ssh_conn_id)
+    with ssh_hook.get_conn() as ssh_client:
+        sftp_client = ssh_client.open_sftp()
+
+        mets_path = f"{BASE_PATH}/mets"
+        api.download_mets(
+            api.download_mets_to_remote, DC_IDENTIFIER, sftp_client, mets_path
+        )
+
+
+def save_alto_files(ssh_conn_id):
+    ssh_hook = SSHHook(ssh_conn_id=ssh_conn_id)
+    with ssh_hook.get_conn() as ssh_client:
+        sftp_client = ssh_client.open_sftp()
+        mets_path = f"{BASE_PATH}/mets"
+        mets_files = sftp_client.listdir(mets_path)
+
+        for filename in mets_files:
+            with sftp_client.file(f"{mets_path}/{filename}", "r") as remote_mets_file:
```

Comment:
The end user does want the METS file :D. 

In the METS/ALTO standard, the ALTO file contains the physical layout segmentation information as well as the OCR of the page, but the METS contains logical segmentation information: newspaper sections, article divisions, paragraph reading order etc. Also, it has scanning metadata on e.g. resolution which has been useful for certain processes where we've needed to calibrate page/image sizes.

Reply:
This might be something that you already know, but the ALTO files described in METS are not actually the ALTOs we will download: pages have been re-OCR'd but the METS files have not been edited. Fetching e.g. the example files from [NLF docs](https://wiki.helsinki.fi/display/Comhis/Interfaces+of+digi.kansalliskirjasto.fi#Interfacesofdigi.kansalliskirjasto.fi-Subcomponentsofindividualbindingavailable) will give us a METS with
```
<fileGrp ID="ALTOGRP" USE="Text">
	<file ID="ALTO00001" CREATED="2022-04-27T06:16:05" MIMETYPE="text/xml" CHECKSUM="f71fead77dcbad4590e9819f41544f55" CHECKSUMTYPE="MD5" SIZE="165706">
<FLocat LOCTYPE="URL" xlink:href="file://./alto/00001.xml"/>
	</file>
	...
</fileGrp>
```
but the checksum of the actual ALTO file does not match:
```
$ md5sum page-1.xml 
44dc84e1f511cf35c84690f88c502ae6  page-1.xml
```

I'ts okay for our download purposes, as new round of OCR won't yield new pages to the binding, but the more fine-grained metadata might be misleading when paired with the new ALTOs.

The same thing goes for the images too: METS files describe the preservation images, but at the moment we can only get the access images, so that metadata is misleading too.

Information about the binding itself should be valid though, so if people want to get e.g. ISSN numbers from the METS files that should be a valid use case and we can provide the files if need be, we'll just need to figure out how to communicate the known discrepancies to users.

Reply:
My understanding is that the new OCR has also been slotted into the old ALTOs with as little intrusion as possible so that on a structural level the METS+ALTO should still continue to be aligned. 

And the METS metadata on the images is also to my understanding transferrable to the access images on a content/descriptive level, even though the *technical* metadata and file paths may be wrong.

There are also all kinds of other problems with the data. Documenting these is crucial, but responsibility for that is split. I think the NLF should document (as text) everything they know, but then after that, for example WP4.3 also relates to this by aiming to provide e.g. statistical overviews of the data that themselves will help to spot anomalies. Downstream enrichers will also discover new problems and we'll need to have a process to include their comments as well.

But from a pragmatic point of view, for now we can assume this documentation happens collaboratively in prose, either on just a GitHub markdown document somewhere, or even in a Google doc or similar. Frex. https://github.com/NatLibFi/Finto-data/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc could be used as one starting template on how to do this. Getting pretty far from the code review here though :D.